### PR TITLE
Add Katacoda embedded tutorials

### LINKED ISSUE #2276: Add Katacoda embedded tutorials
Katacoda embedded tutorials was broken before, see #1874 . Now it's been fixed, so it should be added back.

### PR DESCRIPTION
**Pull request by DaoDaoNoCode**
Closes #2276 . I create a new section in the side nav named `Training`, and there are several pages inside. `About` page contains all the trainingCards of the tutorials, and the other pages are the embedded tutorials. I also hide the title and footer on the tutorial pages.

Thanks @redallen 's help!!

### CODE REVIEW

File: packages/theme-patternfly-org/app.js [hunk 1]
```diff
@@ -7,17 +7,17 @@ import { MDXTemplate } from 'theme-patternfly-org/templates/mdx';
...
 
-const AppRoute = ({ child }) => {
+const AppRoute = ({ child, katacodaLayout }) => {
   const location = useLocation();
   if (typeof window !== 'undefined' && window.ga) {
     window.ga('set', 'page', location.pathname);
     window.ga('send', 'pageview');
   }
-
+ console.log(katacodaLayout)
```
Comment (redallen) [inline packages/theme-patternfly-org/app.js:16]: Remove console.log

**Review by redallen — CHANGES_REQUESTED** [review]
Super close! Can you find a way to remove the padding around `ws-mdx-content`?
image

**Comment by DaoDaoNoCode** [comment]
> Super close! Can you find a way to remove the padding around `ws-mdx-content`?
> image

@redallen Fixed! Thanks for pointing out.

File: packages/v4/src/content/training/tutorials/katacoda-script.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+      const script = document.createElement('script');
+      script.setAttribute('src', '//katacoda.com/embed.js');
+      document.body.appendChild(script);
+      const outterContent = document.getElementsByClassName('ws-mdx-content')[0];
+      outterContent.setAttribute('style', 'padding: 0');
+      const innerContent = document.getElementsByClassName('ws-mdx-content-content')[0];
+      innerContent.setAttribute('style', 'max-width: none');
+      return () => {
+        document.body.removeChild(script);
+      }
```
Comment (redallen) [inline packages/v4/src/content/training/tutorials/katacoda-script.js:15]: Can you wrap this in `if (typeof window !== 'undefined)`? We don't need to this for SSR and it seems to be slowing the build down.

**Review by redallen — CHANGES_REQUESTED** [review]
One last if statement

**Review by redallen — APPROVED** [review]
Looks great to me! If @christiemolloy likes it we can merge.

**Review by christiemolloy — CHANGES_REQUESTED** [review]
Thanks for working on this @DaoDaoNoCode!

Two comments:
Is it possible in the sidebar to split up the tutorials by React and HTML by using the grouping feature of subgroups https://www.patternfly.org/v4/components/navigation#expandable  

When you click on a tutorial from the side nav or the about page is it possible to automatically close the sidenav ?

File: packages/v4/src/content/training/trainingCard/trainingCard.js [hunk 3]
```diff
@@ -50,10 +50,10 @@ export const TrainingCard = ({
       {description}
     </CardBody>
     <CardFooter>
-    <Link to={`https://www.katacoda.com/patternfly/courses/${trainingType}/${katacodaId}`} target="_blank">
+    <Link to={`/training/${katacodaId}`} >
```
Comment (redallen) [inline packages/v4/src/content/training/trainingCard/trainingCard.js:53]: Add `state={{ isKatacoda: true }}`.

File: packages/theme-patternfly-org/app.js [hunk 4]
```diff
@@ -7,39 +7,41 @@ import { MDXTemplate } from 'theme-patternfly-org/templates/mdx';
...
-                sources={sources}
-              />
-            }/>
-        )
-      }
-    </Router>
-  </SideNavLayout>
-);
+const SideNavRouter = () => {
+  const pathName = useLocation().pathname;
```
Comment (redallen) [inline packages/theme-patternfly-org/app.js:25]: I'm not in favor of changing the layout based on path name. Can you implement this using `useLocation().state` instead?

```js
const isNavOpen = !useLocation().state.isKatacoda;
...
<SideNavLayout isNavOpen={isNavOpen}
```

**Review by redallen — CHANGES_REQUESTED** [review]
This will cause too much rerendering, see comments for how to fix it!

File: packages/theme-patternfly-org/routes.js [hunk 5]
```diff
@@ -79,9 +91,18 @@ const getDefaultDesignGuidelines = ({ id, section, slug, title }) => {
   return pageData;
 }
 
-Object.entries(groupedRoutes)
-  .forEach(([_section, ids]) => {
-    Object.values(ids).forEach(pageData => {
+console.log('groupedRoutes', groupedRoutes);
```
Comment (redallen) [inline packages/theme-patternfly-org/routes.js:94]: Remove console.log

File: packages/theme-patternfly-org/scripts/md/parseMD.js [hunk 6]
```diff
@@ -42,13 +42,13 @@ function toReactComponent(mdFilePath, source) {
...
-        frontmatter.section = 'developer-resources';
-      }
       source = frontmatter.source || source;
-      const slug = makeSlug(source, frontmatter.section, frontmatter.id);
+      let slug
+      if (frontmatter.subsection) {
+        slug = makeSlug({ source, section: frontmatter.section, subsection: frontmatter.subsection, id: frontmatter.id });
+      } else {
+        slug = makeSlug({ source, section: frontmatter.section, id: frontmatter.id });
+      }
```
Comment (redallen) [inline packages/theme-patternfly-org/scripts/md/parseMD.js:51]: Change to `const slug = makeSlug({ source, section: frontmatter.section, subsection: frontmatter.subsection, id: frontmatter.id });`. If `frontmatter.subsection` is `undefined` that's okay.

File: packages/v4/src/content/training/trainingCard/trainingCard.js [hunk 7]
```diff
@@ -50,10 +51,10 @@ export const TrainingCard = ({
       {description}
     </CardBody>
     <CardFooter>
-    <Link to={`https://www.katacoda.com/patternfly/courses/${trainingType}/${katacodaId}`} target="_blank">
+    <Link to={`/training/${subsection}/${katacodaId}`} >
```
Comment (redallen) [inline packages/v4/src/content/training/trainingCard/trainingCard.js:54]: I think you should just align `trainingType` and `subsection`. Just change `subsection: react` to `subsection: React charts` for the charts examples.

**Review by redallen — CHANGES_REQUESTED** [review]

**Review by christiemolloy — CHANGES_REQUESTED** [review]
Looking good!
1) I don't think you need "About" as a subsection of "About" could that just be its own section at the top?
2) On second thought I'm thinking could we break up the tutorials by HTML / React / Charts? I think that would help users find the tutorial they're looking for easier.
3) @mcarrano what do you think of the design

**Comment by mcarrano** [comment]
@christiemolloy I'm seeing this for the first time and I agree that the current organization is kind of confusing. I don't think we should be listing the individual tutorials in the leftside Nav, and I like the idea of splitting these into three landing pages (HTML, React, and Charts) as you propose. Not sure we need an "About" section. Is this just intended as a landing page with everything? If so, maybe entitle it as "All" in the navigation?

File: packages/theme-patternfly-org/components/sideNav/sideNav.js [hunk 8]
```diff
@@ -58,6 +58,7 @@ export const SideNav = ({ groupedRoutes = {}, navItems = [] }) => {
                     className="ws-side-nav-group"
                   >
                     {Object.entries(groupedRoutes[section] || {})
+                      .filter(([, { hideNavItem }]) => isUndefined(hideNavItem))
```
Comment (redallen) [inline packages/theme-patternfly-org/components/sideNav/sideNav.js:61]: Change to `!Boolean(hideNavItem)`.

File: packages/theme-patternfly-org/app.js [hunk 9]
```diff
@@ -7,39 +7,45 @@ import { MDXTemplate } from 'theme-patternfly-org/templates/mdx';
...
+  const navOpen = !pathName.startsWith(`${process.env.pathPrefix}/training`)
+                  || pathName === `${process.env.pathPrefix}/training/react`
+                  || pathName === `${process.env.pathPrefix}/training/react-charts`
+                  || pathName === `${process.env.pathPrefix}/training/html`;
```
Comment (redallen) [inline packages/theme-patternfly-org/app.js:29]: Is it possible to look at metadata on `routes` instead? Something like `const navOpen = !routes[pathName].katacodaLayout;`?

**Review by redallen — CHANGES_REQUESTED** [review]
Push again and maybe the build will pass...

File: packages/theme-patternfly-org/app.js (hunk 9 above)
Comment (DaoDaoNoCode) [reply packages/theme-patternfly-org/app.js:29]: It works! Nice suggestion because I am also worrying this sentence is too long.

**Comment by DaoDaoNoCode** [comment]
@mcarrano Could you review the design again?

**Review by mcarrano — APPROVED** [review]
Looks great @DaoDaoNoCode !

**Review by christiemolloy — APPROVED** [review]
Love it!! This looks great thanks @DaoDaoNoCode

**Review by redallen — APPROVED** [review]
LGTM

**Pull request merged**